### DISCUSSION TITLE
Lesson 9 - getting unkown error while solving the Challenge

### DISCUSSION THREAD

**Discussion by abdoMous** [Cyfrin/foundry-full-course-cu#718] [Q&A] [answered]
Hello, I'm encountering an issue with a contract I wrote. I have read through all the hints in the Discussions and tried several suggestions, but I'm still facing the problem. Could someone assist me by providing more insights or hints?

I uploaded the contract to Remix and deployed it on the Sepolia testnet, but I receive an "execution reverted" error.

Here's the contract:

```solidity
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

interface Challenge {
    function solveChallenge(uint256 randomGuess, string memory yourTwitterHandle) external;
}

contract ChallengeSolver {
    
    function solve() public {
        address challengeAddress = 0x33e1fD270599188BB1489a169dF1f0be08b83509;       
        uint256 guess =
            uint256(keccak256(abi.encodePacked(msg.sender, block.prevrandao, block.timestamp))) % 100000;

        Challenge challenge = Challenge(challengeAddress);
        challenge.solveChallenge(guess, "twiterUsername");        
    }
}
```

I appreciate any assistance or feedback!

**Comment by theirrationalone** [comment]
Please search your challenge here on discussion, here is so many hints available. Spoilers are bad. 😉

**Reply by abdoMous** [reply]
I understand. I have read through all the discussions about the challenge, which exactly hint that is available, will help me in this situation, thanks in advanced

**Reply by theirrationalone** [reply]
Why not..... 

Follow [this](https://github.com/Cyfrin/foundry-full-course-f23/discussions/148) thoroughly... 

And Good Luck 🍀 @abdoMous ! 😉

**Reply by abdoMous** [reply]
Hi @theirrationalone, I followed the hint you mentioned, I implemented `onERC721Received` functions, but no result, am I missing something
```solidity
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

// ERC721

import "@openzeppelin/contracts/token/ERC721/IERC721.sol";
import "@openzeppelin/contracts/token/ERC721/IERC721Receiver.sol";


interface Challenge {
    function solveChallenge(uint256 randomGuess, string memory yourTwitterHandle) external;
}


contract ChallengeSolver  is IERC721Receiver  {

    address constant CHALLENGE_ADDRESS = 0x33e1fD270599188BB1489a169dF1f0be08b83509;
    address i_owner;

    constructor () {
        i_owner = msg.sender;
    }

    //////////////////////
    // Public Functions //
    //////////////////////
    
    function solve() public /* returns (bytes4, bool) */ {
        uint256 guess = getTheGess();
        
        // (bool success, bytes memory returnData) = CHALLENGE_ADDRESS.call(abi.encodeWithSelector(getSelector(), guess, "yourTwitterHandle"));
        // return (bytes4(returnData), success);

        Challenge challenge = Challenge(CHALLENGE_ADDRESS);
        challenge.solveChallenge(guess, "yourTwitterHandle");
    }

    // ERC721 Token Receiver Interface implementation
    function onERC721Received(address, address, uint256 tokenId, bytes calldata) external override returns (bytes4) {
        forwardNFT(msg.sender, tokenId);
        return this.onERC721Received.selector; 
    }

    //////////////////////
    // Private Functions //
    //////////////////////

    function getTheGess() private view returns (uint256) {
        return uint256(keccak256(abi.encodePacked(msg.sender, block.prevrandao, block.timestamp))) % 100000;
    }

    function getSelector() private pure returns (bytes4 selector){
        selector = bytes4(keccak256(bytes("solveChallenge(uint256,string)")));
    }

    // This function allows you to safely transfer a NFT to the target address
    function forwardNFT(address nftAddress, uint256 tokenId) private {
        IERC721 nft = IERC721(nftAddress);

        // Ensure that this contract is the owner of the NFT before transferring
        require(nft.ownerOf(tokenId) == address(this), "Not the owner of this token");
        
        nft.safeTransferFrom(address(this), i_owner, tokenId);
    }   
}
```

**Reply by theirrationalone** [reply]
Heyyy @Neftyr would you like to help @abdoMous !

**Reply by Niferu** [reply, marked as answer]
Hi @theirrationalone 
 Sure thing!

@abdoMous the answer is pretty simple, you are passing incorrect answer Sir. Pay extra attention on how it should be calculated, whos calling what etc. just try to simulate this transaction and take a look at challenge contract again.

**Reply by theirrationalone** [reply]
Thanks @Neftyr !, i often scare if i spoil something throughout hints 😅... ;!

**Reply by abdoMous** [reply]
YESSSS!! I've got the NFT! thanks @theirrationalone  and @Neftyr

**Reply by Niferu** [reply]
Good Job @abdoMous !

@theirrationalone  Yeah I was thinking what to write here for like 10 minutes haha to not spoil too much xd

**Reply by theirrationalone** [reply]
Same pinch, that's why i called you 😜😜

**Reply by Ynyesto** [reply]
Hi, man! I've tried something similar to this but it is not working for me... Anyway, I don't really get how your code works, how does onERC721Received get called??

**Reply by theirrationalone** [reply]
Heyy @arynyestos ! You should use console log utility to eve.... Who currently owns the nft, who the previous owner was... And decide yourself to whom you wanna tranfer that nft.... 

Just flood console log inside that event like function... 

Good luck 🍀

**Reply by Ynyesto** [reply]
OK. I've been banging my head against this and at least now I get that onERC721Received is a callback function that gets called by the ERC721 contract upon sending the NFT to my SolveChallenge contract. The thing is I get "Fail with error 'ERC721: transfer to non ERC721Receiver implementer'" as you can see [here](https://sepolia.etherscan.io/tx/0x8b42f136c466ba293a1d881293915887c9cfc016181d1be3b3d31044b536d57b) and I have no idea why, since contract SolveChallenge is IERC721Receiver. I thought the problem could perhaps be that I was calling nft.safeTransferFrom(address(this), owner, tokenId); (that is, from the SolveChallenge contract, to my EOA), but that may not work when sending to a wallet, so I changed it to nft.transferFrom(address(this), owner, tokenId);, but that also doesn't work, since Metamask cannot estimate the gas, so the transaction would revert.  I really don't know what else to try :-( Any ideas/hints? Thanks again!

**Reply by Niferu** [reply]
Hi @arynyestos 

Small hint for you as we do not spoil challenges here. Keep in mind that all ERC721, ERC20 etc. has their own interfaces even if functions looks the same interface is unique per contract :)

**Reply by Ynyesto** [reply]
Hey, man! Thanks for your help! That wasn't the issue, I was already using the right interfaces, but I was passing the wrong address for the NFT contract 🤦‍♂️🤦‍♂️🤦‍♂️🤦‍♂️

**Reply by Niferu** [reply]
Hahaha @arynyestos 

That happens also :D , I'm glad you have solved the issue and challenge, cheers! ;)

**Reply by Ynyesto** [reply]
Thanks a lot!!
